segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/MetadataStore.java

```diff
@@ -373,17 +388,23 @@
 
     /**
      * Attempts to map a Segment to an Id, by first trying to retrieve an existing id, and, should that not exist,
-     * assign a new one.
+     * assign a new one. If the operation failed, either synchronously, or asynchronously, the segment assignment will be
+     * failed with the causing exception.
      *
      * @param segmentName The name of the Segment to assign id for.
      * @param timeout     Timeout for the operation.
      */
     private void assignSegmentId(String segmentName, Duration timeout) {
-        TimeoutTimer timer = new TimeoutTimer(timeout);
-        Futures.exceptionListener(
-                getSegmentInfoInternal(segmentName, timer.getRemaining())
-                        .thenComposeAsync(si -> submitAssignmentWithRetry(SegmentInfo.deserialize(si), timer.getRemaining()), this.executor),
-                ex -> failAssignment(segmentName, ex));
+        try {
```

Comment:
It seems like there needs to be a check if the Store is closed here.

Reply:
Either `getSegmentInfoInternal` or `submitAssignment` will throw that. The former will resort to the `Storage` adapter which will be closed (if the container is closed) and the latter will run into the `OperationProcessor`, which is the first thing that gets shut down.

The latter is exactly what triggered this problem in the first place, and is tested via the new unit test added.